resources/net.go

```diff
@@ -0,0 +1,804 @@
+		return false, errwrap.Wrapf(err, "%s", slurp)
+	}
+	// XXX: wait for state?
+
+	return false, nil
+}
+
+// fileCheckApply checks and maintains the systemd-networkd unit file contents.
+// TODO: check for conflicting unit files and think about unit priority
+func (obj *NetRes) fileCheckApply(apply bool) (checkOK bool, _ error) {
+	// check if the unit file exists
+	_, err := os.Stat(obj.unitFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return false, errwrap.Wrapf(err, "error checking file")
+		}
+	}
+	// build the unit file contents from the definition
+	contents := obj.unitFileContents()
+	// check the file contents
+	if err == nil {
+		unitFile, err := ioutil.ReadFile(obj.unitFilePath)
+		if err != nil {
+			return false, errwrap.Wrapf(err, "error reading file")
+		}
+		// return if the file is good
+		if bytes.Equal(unitFile, contents) {
+			return true, nil
+		}
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	log.Printf("%s: fileCheckApply(%t)", obj, apply)
+
+	// write the file
+	if err := ioutil.WriteFile(obj.unitFilePath, contents, 0644); err != nil {
+		return false, errwrap.Wrapf(err, "error writing configuration file")
+	}
+	return false, nil
+}
+
+// ipCheckApply checks if the interface has the correct addresses and restarts
+// the networkd service if the addresses are inconsistent.
+func (obj *NetRes) ipCheckApply(apply bool) (checkOK bool, err error) {
+	var addrStrings []string
+	addrOK := true
+
+	if obj.Addrs == nil {
+		return true, nil
+	}
+
+	// if the interface is configuring, wait for the next event
+	// create a bus object pointing to the adapter
+	busObj := obj.gdbus.Object(dBusNetworkdInterface, obj.ifacePath)
+	// get the object's operational state
+	call := busObj.Call(dBusPropertiesGet, 0, dBusNetworkdLinkInterface, adminState)
+	if call.Err != nil {
+		return false, call.Err
+	}
+	// check the operational state
+	as, ok := call.Body[0].(dbus.Variant)
+	if !ok {
+		return false, fmt.Errorf("error casting opstate")
+	}
+	if as == dbus.MakeVariant(adminStateConfiguring) {
+		return false, nil
+	}
+
+	// find the specified interface and get a list of addresses
+	iface, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error getting addrs from interface: %s", obj.GetName())
+	}
+	// if the address is not a link local address, add it to the list
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			return false, errwrap.Wrapf(err, "error parsing interface addr: %s", addr)
+		}
+		// XXX: ignore multicast too?
+		if !ip.IsLinkLocalUnicast() {
+			addrStrings = append(addrStrings, addr.String())
+		}
+	}
+
+	// compare defined addrs with actual addrs
+	objAddrs := obj.Addrs
+	sort.Strings(objAddrs)
+	sort.Strings(addrStrings)
+	if len(objAddrs) == len(addrStrings) {
```

Comment:
this sort of deep nesting either needs inline-ing, or made into a helper function please